# KNN Aufgabe mit Iris Dataset

a) Load the Iris dataset using load_iris() and split into training/testing sets.
Scale the data using StandardScaler.
Train a KNeighborsClassifier, experiment with n_neighbors and p, and calculate the test score.

b) Train a GaussianNB (Naive Bayes) and LogisticRegression on the scaled training data and compute their test scores.

c) Use cross_val_score for 5-fold cross-validation on each model to compute mean and standard deviation of their scores, comparing their performance across folds.

In [ ]:
# Step 1: Load Iris dataset
from sklearn.datasets import load_iris
iris = load_iris()

# Load and split the data
from sklearn.model_selection import train_test_split
X, y = iris.data, iris.target  # Load features and target
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)


In [4]:
# Scale the data using StandardScaler
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


In [12]:
# a) K-Nearest Neighbors
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=3, p=2)  # Experiment with n_neighbors and p
knn.fit(X_train_scaled, y_train)
knn_score = knn.score(X_test_scaled, y_test)
print(f"KNN Test Score: {knn_score:.4f}")

KNN Test Score: 1.0000


In [13]:
# b) Naive Bayes and Logistic Regression
from sklearn.naive_bayes import GaussianNB
gnb = GaussianNB()
gnb.fit(X_train_scaled, y_train)
gnb_score = gnb.score(X_test_scaled, y_test)
print(f"Naive Bayes Test Score: {gnb_score:.4f}")

Naive Bayes Test Score: 0.9778


In [14]:
# b) Naive Bayes and Logistic Regression
from sklearn.linear_model import LogisticRegression
log_reg = LogisticRegression()
log_reg.fit(X_train_scaled, y_train)
log_reg_score = log_reg.score(X_test_scaled, y_test)
print(f"Logistic Regression Test Score: {log_reg_score:.4f}")

Logistic Regression Test Score: 1.0000


In [15]:
from sklearn.model_selection import cross_val_score
knn_cv_scores = cross_val_score(knn, X_train_scaled, y_train, cv=5)
gnb_cv_scores = cross_val_score(gnb, X_train_scaled, y_train, cv=5)
log_reg_cv_scores = cross_val_score(log_reg, X_train_scaled, y_train, cv=5)


In [16]:
# Compare cross-validation results
print("\nCross-Validation Results:")
print(f"KNN: Mean={knn_cv_scores.mean():.4f}, Std={knn_cv_scores.std():.4f}")
print(f"Naive Bayes: Mean={gnb_cv_scores.mean():.4f}, Std={gnb_cv_scores.std():.4f}")
print(f"Logistic Regression: Mean={log_reg_cv_scores.mean():.4f}, Std={log_reg_cv_scores.std():.4f}")


Cross-Validation Results:
KNN: Mean=0.9429, Std=0.0467
Naive Bayes: Mean=0.9333, Std=0.0381
Logistic Regression: Mean=0.9429, Std=0.0356


The mean score represents the average accuracy of the model across the 5 cross-validation folds. Higher mean scores indicate better average performance. All three models perform well overall, with accuracies above 93%.

The standard deviation measures how much the scores vary across different folds. Lower standard deviation means more consistent performance across folds, suggesting the model generalizes well. Best Average Accuracy: KNN and Logistic Regression tie for the highest mean accuracy.